internal/impl/postgresql/pglogicalstream/logical_stream.go

```diff
@@ -96,85 +111,120 @@ func NewPgStream(config Config) (*Stream, error) {
-		tableNames[i] = fmt.Sprintf("%s.%s", config.DbSchema, table)
+		tableNames[i] = fmt.Sprintf("%s.%s", config.DBSchema, table)
 	}
 
-	tablesSchemaFilter := fmt.Sprintf("FOR TABLE %s", strings.Join(tableNames, ","))
-	stream.logger.Infof("Create publication for table schemas with query %s", fmt.Sprintf("CREATE PUBLICATION pglog_stream_%s %s;", config.ReplicationSlotName, tablesSchemaFilter))
-	result = stream.pgConn.Exec(context.Background(), fmt.Sprintf("CREATE PUBLICATION pglog_stream_%s %s;", config.ReplicationSlotName, tablesSchemaFilter))
-	_, err = result.ReadAll()
+	var version int
+	version, err = getPostgresVersion(*cfg)
 	if err != nil {
-		panic(fmt.Errorf("create publication error %w", err)) // TODO
+		return nil, err
 	}
-	stream.logger.Infof("Created Postgresql publication %v %v", "publication_name", config.ReplicationSlotName)
 
-	sysident, err := pglogrepl.IdentifySystem(context.Background(), stream.pgConn)
+	snapshotter, err := NewSnapshotter(stream.dbConfig, stream.logger, version)
 	if err != nil {
-		panic(fmt.Errorf("failed to identify the system %w", err)) // TODO
+		stream.logger.Errorf("Failed to open SQL connection to prepare snapshot: %v", err.Error())
+		if err = stream.cleanUpOnFailure(); err != nil {
+			stream.logger.Errorf("Failed to clean up resources on accident: %v", err.Error())
+		}
+
+		os.Exit(1)
+	}
+	stream.snapshotter = snapshotter
+
+	var pluginArguments = []string{}
+	if stream.decodingPlugin == "pgoutput" {
+		pluginArguments = []string{
+			"proto_version '1'",
+			fmt.Sprintf("publication_names 'pglog_stream_%s'", config.ReplicationSlotName),
+		}
+
+		if version > 14 {
+			pluginArguments = append(pluginArguments, "messages 'true'")
+		}
+	}
+
+	if stream.decodingPlugin == "wal2json" {
+		tablesFilterRule := strings.Join(tableNames, ", ")
+		pluginArguments = []string{
+			"\"pretty-print\" 'true'",
+			"\"add-tables\"" + " " + fmt.Sprintf("'%s'", tablesFilterRule),
+		}
+	}
+
+	stream.decodingPluginArguments = pluginArguments
+
+	// create snapshot transaction before creating a slot for older PostgreSQL versions to ensure consistency
+
+	pubName := "pglog_stream_" + config.ReplicationSlotName
+	fmt.Println("Creating publication", pubName, "for tables", tableNames)
+	if err = CreatePublication(context.Background(), stream.pgConn, pubName, tableNames, true); err != nil {
+		return nil, err
 	}
 
-	stream.logger.Infof("System identification result SystemID: %v Timeline: %v XLogPos: %v DBName: %v", sysident.SystemID, sysident.Timeline, sysident.XLogPos, sysident.DBName)
+	sysident, err := IdentifySystem(context.Background(), stream.pgConn)
+	if err != nil {
+		return nil, err
+	}
 
 	var freshlyCreatedSlot = false
 	var confirmedLSNFromDB string
 	// check is replication slot exist to get last restart SLN
 	connExecResult := stream.pgConn.Exec(context.TODO(), fmt.Sprintf("SELECT confirmed_flush_lsn FROM pg_replication_slots WHERE slot_name = '%s'", config.ReplicationSlotName))
```

Comment:
Can we pass in a context here? There should be one from the input, same for the ones above.